### DISCUSSION TITLE
[categorial labeling, trame] Workaround for lost legends and axes labels when exported to html / trame

### DISCUSSION THREAD

**Discussion by MarkusZimmerDLR** [pyvista/pyvista#5249] [Q&A]
Dear pyvista folks,

I am currently struggling with the trame export of pyvista plotters.

More specifically, I am trying to mark and label specific parts of the mesh, either using labels and legends, or labelled axes and cell data.

Legends are lost after html export. Axes labels are converted to integers.
Any idea on how to generate a categorial labelling.

Thanks!

Quick example of the issue:

```python
import pyvista as pv
import numpy as np
s = pv.Sphere()
a = np.arange(s.n_cells, dtype=object)
a[:s.n_cells//2] = 'spam'
a[s.n_cells//2:] = 'egg'
s.cell_data['test'] = a
p = pv.Plotter()
p.add_mesh(s, scalars='test')
p.export_html('test.html')
p.screenshot('test.png')
```
Result for the png:
![test](https://github.com/pyvista/pyvista/assets/77002394/5a1de292-9214-435b-b599-7c7e431730f2)

Result for html (converted to png for upload):

![grafik](https://github.com/pyvista/pyvista/assets/77002394/dd125b59-f55a-495e-96a1-02d2b0b78aec)

**Comment by MarkusZimmerDLR** [comment]
```
--------------------------------------------------------------------------------
  Date: Fri Dec 01 11:21:55 2023 CET

                  OS : Linux
              CPU(s) : 128
             Machine : x86_64
        Architecture : 64bit
                 RAM : 251.7 GiB
         Environment : IPython
         File system : xfs
          GPU Vendor : Mesa
        GPU Renderer : llvmpipe (LLVM 17.0.1, 256 bits)
         GPU Version : 4.5 (Core Profile) Mesa 23.2.1
    MathText Support : False

  Python 3.12.0 | packaged by conda-forge | (main, Oct  3 2023, 08:43:22) [GCC
  12.3.0]

             pyvista : 0.42.3
                 vtk : 9.2.6
               numpy : 1.26.2
          matplotlib : 3.8.2
              scooby : 0.9.2
               pooch : 1.8.0
              pillow : 10.1.0
             imageio : 2.33.0
           pyvistaqt : 0.11.0
               PyQt5 : 5.15.10
             IPython : 8.18.1
          ipywidgets : 8.1.1
               scipy : 1.11.4
                tqdm : 4.66.1
              meshio : 5.3.4
               trame : 3.3.0
        trame_client : 2.12.7
        trame_server : 2.12.1
           trame_vtk : 2.6.2
       trame_vuetify : 2.3.1
jupyter_server_proxy : 4.1.0
        nest_asyncio : 1.5.8
--------------------------------------------------------------------------------
```

**Comment by banesullivan** [comment]
This is unfortunately a limitation of VTK.js and [trame-vtk](https://github.com/Kitware/trame-vtk) when using the client rendering modes

**Reply by MarkusZimmerDLR** [reply]
I am seriously interested in a solution. 
If there is no workaround to get some kind of categorical labeling, please point in the right direction, so I can create a feature request in the correct repository (vtk.js / trame-vtk).
I am pretty sure I am not the only one with a use case, and this will come up more often in the future.

**Reply by jourdain** [reply]
The first step would be to add support in vtk.js. Probably in the [ScalarBarActor](https://github.com/Kitware/vtk-js/tree/master/Sources/Rendering/Core/ScalarBarActor) and its implementations [WebGL](https://github.com/Kitware/vtk-js/blob/master/Sources/Rendering/OpenGL/ScalarBarActor/index.js)/[WebGPU](https://github.com/Kitware/vtk-js/blob/master/Sources/Rendering/WebGPU/ScalarBarActor/index.js).
Once working on the vtk.js side, we may need to adjust the scene serializer, but it should be pretty straightforward if we have "parity" between the C++ server and the JS client.

**Reply by MarkusZimmerDLR** [reply]
OK, so I will go ahead and create an FR in the vtk.js repository. However, I am afraid I won't be of mutch help. JS and c++ are not on my strong side.

As mentioned in the first post, legends are also not working yet. But as these are implemented as widgets and widgets are also generally not supported in vtk.js, I do not expect this to be happening any time soon.

However, I want to thank you for the awesome framework you guys have created. It is a blast to work with!

**Comment by MarkusZimmerDLR** [comment]
I figured. But can you think of any workaround which will enable me to label something in the plot?

**Reply by jourdain** [reply]
If you are using trame, you can build such annotation as an overlay (or side) in plain HTML

**Reply by jourdain** [reply]
FYI, with trame we are currently working on a VTK/WASM implementation to get the full parity with local rendering.

**Reply by MarkusZimmerDLR** [reply]
That is a lot to put in one sentence ;) Can you elaborate a little more on what that means for this case? Does that mean you will ditch the vtk.js framework?

**Reply by jourdain** [reply]
vtk-js will still exist has it provide other benefits than a WASM implementation. But for our usage in trame where we want to mirror a VTK scene that exist on the server side but rendering it on the client side, WASM will give us the full parity that we are currently missing with vtk.js.

**Reply by MarkusZimmerDLR** [reply]
This. Sounds. Awesome! 
Can you guide me to a related issue / repo where I can keep up with updates? I cannot find anything under the pyvista namespace.

**Reply by jourdain** [reply]
It is currently in development (/private => Kitware/trame-vtklocal) but aimed to be available by the next VTK 9.4 release (April) .
We are still working on [C++](https://gitlab.kitware.com/vtk/vtk/-/merge_requests/10898) and CI.

**Reply by MarkusZimmerDLR** [reply]
Any news regarding this? The linked MR seems a little dead to me.

**Reply by MarkusZimmerDLR** [reply]
With the current releases, the example from the initial post, exported to html, results in a white screen without anything being rendered.

This is not caused by the categorial labelling. The Problem occurs as soon as the `scalars="xyz"` kwarg is added to `p.add_mesh`.
The exported png is fine. It does not depend on the data type stored in the plotted array.

```
--------------------------------------------------------------------------------
  Date: Wed Sep 11 18:19:14 2024 CEST

                  OS : Linux (Rocky Linux 8.9)
              CPU(s) : 128
             Machine : x86_64
        Architecture : 64bit
                 RAM : 251.7 GiB
         Environment : IPython
         File system : xfs
          GPU Vendor : Mesa
        GPU Renderer : llvmpipe (LLVM 18.1.8, 256 bits)
         GPU Version : 4.5 (Core Profile) Mesa 24.1.5
    MathText Support : True

  Python 3.12.5 | packaged by conda-forge | (main, Aug  8 2024, 18:36:51) [GCC
  12.4.0]

             pyvista : 0.44.1
                 vtk : 9.3.1
               numpy : 2.1.0
          matplotlib : 3.9.2
              scooby : 0.10.0
               pooch : 1.8.2
              pillow : 10.4.0
             imageio : 2.35.1
           pyvistaqt : 0.11.1
               PyQt5 : 5.15.11
             IPython : 8.26.0
          ipywidgets : 8.1.5
               scipy : 1.14.1
                tqdm : 4.66.5
              meshio : 5.3.5
               trame : 3.6.3
        trame_client : 3.2.5
        trame_server : 3.1.0
           trame_vtk : 2.8.10
       trame_vuetify : 2.6.2
jupyter_server_proxy : 4.3.0
        nest_asyncio : 1.6.0
--------------------------------------------------------------------------------
```

**Reply by MarkusZimmerDLR** [reply]
@jourdain

**Reply by jourdain** [reply]
trame-vtklocal and vtk 9.4 provides a solution but that also mean that PyVista and VTK 9.4+ needs to play well together. And we are still working on that.

**Reply by Dan2712** [reply]
> trame-vtklocal and vtk 9.4 provides a solution but that also mean that PyVista and VTK 9.4+ needs to play well together. And we are still working on that.

Really looking forward this 👍 Is there any estimated time for this ❤️

**Reply by jourdain** [reply]
I think PyVista needs 9.4.1 which is supposed to happen in begining on January. After that, I'm not sure about the PyVista side to be fully ok with 9.4.1. Also, someone will have to take on the trame-vtklocal into PyVista unless you do it on your side/app.

**Comment by kevinalexandr19** [comment]
I have the same problem when I try to export a 3D visualization with a title, the title doesn't appear in the HTML export.

**Reply by jourdain** [reply]
vtk.js does not support text labels hence the issue.
